src/webviews.ts

```diff
@@ -1,43 +1,26 @@
-import {
-    merge,
-    map,
-    take,
-    toArray,
-    Observable,
-    of
-} from 'rxjs';
-import type { Webview } from 'vscode';
+import { Observable } from 'rxjs';
 import type { Provider } from './types';
 import BaseChannel from './channel';
+import { Bus } from './tangle';
+import { Webview } from 'vscode';
 
 /**
- * In VSCode web views are created when opened by the extension
+ * In VS Code web views are created when opened by the extension
  * therefor we don't always know when this instance is available.
  * RxJS helps here to initiate the channel only once that happened.
  */
-type WebviewProvider = Observable<Webview> | Webview;
 
-export default class WebViewChannel<T> extends BaseChannel<WebviewProvider, T> {
-    public providers: Observable<Webview>[] = [];
-
-    register(providers: WebviewProvider[]) {
-        const observableProvider: Observable<Webview>[] = providers.map((p) => {
-            const panel = p as Webview;
-            return typeof panel.html === 'string' ? of(panel) : (p as Observable<Webview>);
-        });
-
-        this.providers.push(...observableProvider);
-        return merge(...observableProvider).pipe(
-            take(providers.length),
-            toArray(),
-            map((ps: Webview[]) => ps.map((p) => (<Provider>{
-                onMessage: p.onDidReceiveMessage.bind(p),
-                postMessage: p.postMessage.bind(p),
-            }))),
-            map(this._initiateBus.bind(this))
-        );
+export default class WebViewChannel<T> extends BaseChannel<Webview, T> {
+    register(providers: Observable<Webview>[]): Observable<Bus<T>> ;
+    register(providers: Webview[]): Observable<Bus<T>> ;
+    register(providers: any): Observable<Bus<T>> {
+        return this._register(providers, (p) => (<Provider>{
+            onMessage: p.onDidReceiveMessage.bind(p),
+            postMessage: p.postMessage.bind(p),
+        }));
     }
 
+    // TODO: perhaps type union here? since technically DOM-based envs don't have Webview
```

Comment:
What do you exactly mean by "type union here"? I'ld suggest maybe check if we are in a vscode environment and throw an error if not.

Reply:
What I mean is we can probably figure out a expected type here to satisfy this being called running inside of a DOM vs Node context to avoid something like this:

https://github.com/stateful/vscode-awesome-ux/blob/main/src/components/settings.ts#L49

Eg a type alias for https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/vscode-webview/index.d.ts#L12

Then we should be able to do `attach(webview: Webview | WebviewApi)` and avoid casting on the consumer side.

Reply:
I believe this is actually what we want @christian-bromann: https://github.com/stateful/tangle/commit/dedd6741ae81c19196948050c4cafc73dc0a9175. Don't need a union since the DOM will just have `window` available.